neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/bolt/impl/BoltNeo4jSchemaDefiner.java

```diff
@@ -0,0 +1,118 @@
+import org.hibernate.ogm.datastore.neo4j.impl.BaseNeo4jSchemaDefiner;
+import org.hibernate.ogm.datastore.neo4j.logging.impl.Log;
+import org.hibernate.ogm.datastore.neo4j.logging.impl.LoggerFactory;
+import org.hibernate.ogm.datastore.neo4j.remote.bolt.dialect.impl.BoltNeo4jSequenceGenerator;
+import org.hibernate.ogm.datastore.spi.DatastoreProvider;
+import org.hibernate.ogm.model.key.spi.IdSourceKeyMetadata;
+import org.hibernate.tool.hbm2ddl.UniqueConstraintSchemaUpdateStrategy;
+import org.neo4j.driver.v1.Driver;
+import org.neo4j.driver.v1.Session;
+import org.neo4j.driver.v1.Statement;
+import org.neo4j.driver.v1.StatementResult;
+import org.neo4j.driver.v1.Transaction;
+import org.neo4j.driver.v1.exceptions.ClientException;
+import org.neo4j.driver.v1.util.Resource;
+
+/**
+ * Initialize the schema for the Neo4j database:
+ * <ol>
+ * <li>create sequences;</li>
+ * <li>create unique constraints on identifiers, natural ids and unique columns</li>
+ * </ol>
+ * <p>
+ * Note that unique constraints involving multiple columns won't be applied because Neo4j does not support it.
+ * <p>
+ * The creation of unique constraints can be skipped setting the property
+ * {@link Environment#UNIQUE_CONSTRAINT_SCHEMA_UPDATE_STRATEGY} to the value
+ * {@link UniqueConstraintSchemaUpdateStrategy#SKIP}. Because in Neo4j unique constraints don't have a name, setting the
+ * value to {@link UniqueConstraintSchemaUpdateStrategy#RECREATE_QUIETLY} or
+ * {@link UniqueConstraintSchemaUpdateStrategy#DROP_RECREATE_QUIETLY} will have the same effect: keep the existing
+ * constraints and create the missing one.
+ *
+ * @author Davide D'Alto
+ * @author Gunnar Morling
+ */
+public class BoltNeo4jSchemaDefiner extends BaseNeo4jSchemaDefiner {
+
+	private static final Log log = LoggerFactory.getLogger();
+
+	@Override
+	protected void createSequences(List<Sequence> sequences, Set<IdSourceKeyMetadata> allIdSourceKeyMetadata, DatastoreProvider provider) {
+		BoltNeo4jDatastoreProvider boltProvider = (BoltNeo4jDatastoreProvider) provider;
+		BoltNeo4jSequenceGenerator sequenceGenerator = boltProvider.getSequenceGenerator();
+		sequenceGenerator.createSequences( sequences, allIdSourceKeyMetadata );
+	}
+
+	@Override
+	protected void createUniqueConstraintsIfMissing(DatastoreProvider provider, List<UniqueConstraintDetails> constraints) {
+		List<Statement> statements = new ArrayList<>();
+		for ( UniqueConstraintDetails constraint : constraints ) {
+			log.tracef( "Creating unique constraint for nodes labeled as %1$s on property %2$s", constraint.getLabel(), constraint.getProperty() );
+			StringBuilder queryBuilder = new StringBuilder( "CREATE CONSTRAINT ON (n:" );
+			escapeIdentifier( queryBuilder, constraint.getLabel().name() );
+			queryBuilder.append( ") ASSERT n." );
+			escapeIdentifier( queryBuilder, constraint.getProperty() );
+			queryBuilder.append( " IS UNIQUE" );
+			statements.add( new Statement( queryBuilder.toString() ) );
+		}
+
```

Comment:
Bolt and Neo4j could probably share most of this implementation but I suppose you thought about it and preferred this approach?